Cocoa SharedMemory::Handle should use MachSendRight

kkinnunen-apple opened the pull request:
#### a5a7844e79bd390f4994553e4c5e82cd40e37b77
<pre>
Cocoa SharedMemory::Handle should use MachSendRight
<a href="https://bugs.webkit.org/show_bug.cgi?id=244600">https://bugs.webkit.org/show_bug.cgi?id=244600</a>
rdar://problem/99379026

Reviewed by Antti Koivisto.

* Source/WebKit/Platform/SharedMemory.h:
* Source/WebKit/Platform/cocoa/SharedMemoryCocoa.cpp:
(WebKit::SharedMemory::Handle::takeOwnershipOfMemory const):
(WebKit::SharedMemory::Handle::setOwnershipOfMemory const):
(WebKit::SharedMemory::Handle::isNull const):
(WebKit::SharedMemory::Handle::clear):
(WebKit::SharedMemory::Handle::encode const):
(WebKit::SharedMemory::Handle::decode):
(WebKit::SharedMemory::allocate):
(WebKit::SharedMemory::wrapMap):
(WebKit::SharedMemory::map):
(WebKit::SharedMemory::~SharedMemory):
(WebKit::SharedMemory::createHandle):
(WebKit::SharedMemory::createSendRight const):
(WebKit::SharedMemory::Handle::Handle): Deleted.
(WebKit::SharedMemory::Handle::operator=): Deleted.
(WebKit::SharedMemory::Handle::~Handle): Deleted.

Canonical link: <a href="https://commits.webkit.org/254037@main">https://commits.webkit.org/254037@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/08ee468c1b1263e2c1c71d31d1502c04efaaca03

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/87720 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/69/builds/31806 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/18456 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/96915 "Built successfully") | 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/64/builds/30164 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/26265 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/79818 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/91661 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/93339 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/68/builds/24365 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/74467 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/24364 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/79327 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/79510 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/67213 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/27858 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/13331 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/27824 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/14346 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/2838 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/29519 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/37253 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/29417 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/33622 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

mattwoodrow commented on Source/WebKit/Platform/SharedMemory.h:
```diff
@@ -99,11 +100,11 @@ class SharedMemory : public ThreadSafeRefCounted<SharedMemory> {
         mutable IPC::Attachment m_attachment;
 #elif OS(DARWIN)
-        mutable mach_port_t m_port { MACH_PORT_NULL };
+        mutable MachSendRight m_handle;
```
Call this one m_sendRight as well for consistency?

kkinnunen-apple replied:
The eventual idea is to call all the other ports' handles m_handle and move the decode/encode code to SharedMemory.cpp to be shared. The textual code will be the same for all.

mattwoodrow commented on Source/WebKit/Platform/cocoa/SharedMemoryCocoa.cpp:
```diff
@@ -130,28 +114,23 @@ void SharedMemory::Handle::setOwnershipOfMemory(const WebCore::ProcessIdentity&
     encoder << static_cast<uint64_t>(m_size);
-    encoder << MachSendRight::adopt(m_port);
-    m_port = MACH_PORT_NULL;
+    encoder << WTFMove(m_handle); // FIXME: add rvalue encode.
```
Is WTFMove doing anything here, or it just invoking the copy operator of encoder? This is const function, so we should be copying the send right into the encoder.

Seems like it'd be clearer to just not have WTFMove.

kkinnunen-apple replied:
We should be copying, but as the added move and the previous code mentions, the intention is to move the mutable m_handle into the encoder, similar what was before.

mattwoodrow reviewed (approved):
Looks great to me!

I'm not a WebKit reviewer though, so you'll need someone else to approve it for merge-queue to work I believe.

kkinnunen-apple committed:
Cocoa SharedMemory::Handle should use MachSendRight
https://bugs.webkit.org/show_bug.cgi?id=244600
rdar://problem/99379026

Reviewed by Antti Koivisto.

* Source/WebKit/Platform/SharedMemory.h:
* Source/WebKit/Platform/cocoa/SharedMemoryCocoa.cpp:
(WebKit::SharedMemory::Handle::takeOwnershipOfMemory const):
(WebKit::SharedMemory::Handle::setOwnershipOfMemory const):
(WebKit::SharedMemory::Handle::isNull const):
(WebKit::SharedMemory::Handle::clear):
(WebKit::SharedMemory::Handle::encode const):
(WebKit::SharedMemory::Handle::decode):
(WebKit::SharedMemory::allocate):
(WebKit::SharedMemory::wrapMap):
(WebKit::SharedMemory::map):
(WebKit::SharedMemory::~SharedMemory):
(WebKit::SharedMemory::createHandle):
(WebKit::SharedMemory::createSendRight const):
(WebKit::SharedMemory::Handle::Handle): Deleted.
(WebKit::SharedMemory::Handle::operator=): Deleted.
(WebKit::SharedMemory::Handle::~Handle): Deleted.

Canonical link: https://commits.webkit.org/254037@main
